text/287-cli-deprecation-warnings.md

```diff
@@ -0,0 +1,224 @@
+
+We can use [TypeScript decorators](https://www.typescriptlang.org/docs/handbook/decorators.html)
+to enhance the runtime JavaScript code with our custom logic.
+We can add an `awslint` rule that enforces that every element with the `@deprecated`
+JSDoc tag also needs to have the `@deprecated` decorator present.
+While decorators are still considered an experimental feature,
+our init templates and JSII enable them in `tsconfig.json`,
+which means the vast majority of our customers will have them enabled as well.
+
+The main advantage of this solution is that changes are needed only in the CDK project.
+
+Disadvantages of this solution:
+
+1. It won't be possible to reliably get the module the deprecated element belongs to (only the name of the element) --
+  the decorator is simply a JavaScript function defined in `@aws-cdk/core`,
+  and doesn't have (easy) access to the information on what module a given object it's called on belongs to.
+2. TypeScript does not allow decorators on interfaces (neither on the interface directly, nor on any of its properties).
+  This means we will not be able to register warnings for structs, or struct properties.
+3. We won't be able to register warnings for accessing deprecated static constants
+   (this includes enums).
+
+#### 3. Rely on language-specific deprecation warning mechanisms
+
+JSII emits deprecated elements with the correct language-specific mechanisms in place
+(for example, the `@Deprecated` annotation for Java).
+We could simply piggyback on those,
+instead of writing our own.
+
+Advantages of this solution:
+
+1. Minimal development effort on our side.
+
+Disadvantages of this solution:
+
+1. Just deprecated warnings in the IDE can be missed by developers
+  (especially as more APIs get deprecated).
+2. The experience with TypeScript properties is currently pretty bad
+  (the property is struck through when selecting it,
+  but not after that!).
+3. There is nothing out of the box for this in Python
+  (we would have to write our own decorator).
+
+#### 4. Parse the customer's code
+
+We could create a parser and type-checker for each language the CDK supports,
+and use that to discover any cases of using deprecated elements.
+
+Advantages of this solution:
+
+1. Allows us to discover some tricky deprecated elements,
+  like enum values, or deprecated types used only in type declarations,
+  that will be difficult to do with only runtime reporting.
+2. Would prevent reporting warnings for deprecated usages outside the customer's code
+  (for example, in our own libraries).
+
+Disadvantages of this solution:
+
+1. We would have to write a separate parser and type-checker for each language supported by the CDK.
+2. This additional parsing could have an adverse impact on the performance of CDK commands.
+3. It's not obvious this analysis can even be performed at all in the case of dynamically-typed languages like JavaScript or Python
+
+### What is the high level implementation plan?
```

Comment:
For the sake of planning, can you also attach high level effort estimates for each item.

Reply:
Done.